api/libproject/devstart.rb

```diff
@@ -1605,6 +1605,66 @@ def export_workspace_data(cmd_name, *args)
     :description => "Export workspace data to CSV.\n",
     :fn => ->(*args) {export_workspace_data("export-workspace-data", *args)}
 })
+def generate_impersonated_user_token(cmd_name, *args)
+  common = Common.new
+  ensure_docker cmd_name, args
+
+  op = WbOptionsParser.new(cmd_name, args)
+  op.add_typed_option(
+      "--output-token-filename [output-token-filename]",
+      String,
+      ->(opts, v) { opts.output_token_filename = v},
+      "Path to an output file for the generated token")
+  op.add_typed_option(
+      "--impersonated-username [impersonated-username]",
+      String,
+      ->(opts, v) { opts.impersonated_username = v},
+      "AoU researcher email to impersonate, e.g. calbach@fake-research-aou.org")
+  op.add_validator ->(opts) { raise ArgumentError unless (opts.output_token_filename and opts.impersonated_username)}
+  op.parse.validate
+
+  project_id = nil
+  ENVIRONMENTS.each_key do |project|
+    if project == "local"
+      next
+    end
+
+    config = get_config(project)
+    if op.opts.impersonated_username.end_with?("@" + config["googleDirectoryService"]["gSuiteDomain"])
+      project_id = project
+      break
+    end
+  end
+  if project_id.nil?
+    common.error "invalid domain for given user #{op.opts.impersonated_username} - target must be an AoU research domain email"
+    raise ArgumentError
+  end
+
+  if ["all-of-us-rw-prod", "all-of-us-rw-preprod"].include? project_id
+    get_user_confirmation(
+      "Using impersonation in a production environment is highly discouraged, " +
+      "and should only be considered in a break-glass scenario. Check with the " +
+      "Workbench team that all other options have been exhausted before " +
+      "continuing. Continue?")
+  end
+
+  flags = ([
+      ["--output-token-filename", op.opts.output_token_filename],
+      ["--impersonated-username", op.opts.impersonated_username],
+      ["--project-id", project_id]
+  ]).map { |kv| "#{kv[0]}=#{kv[1]}" }
+  flags.map! { |f| "'#{f}'" }
+
+  common.run_inline %W{
```

Comment:
I don't think I see a step that pulls down the correct sa key. Will this work if the sa-key is missing or if I want to switch environments?

Reply:
Great catch. my goal was actually to just use the `pmi-ops` creds for this interaction, then exchange from there to the admin account. However, I forgot that as soon as we enter docker we are coercing our credentials to match whatever is sitting in the ADC GOOGLE_APPLICATION_CREDENTIALS file.